aptos-move/mvhashmap/src/versioned_code.rs

```diff
@@ -0,0 +1,214 @@
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::types::{MVCodeError, MVCodeOutput, TxnIndex};
+use aptos_crypto::hash::{DefaultHasher, HashValue};
+use aptos_types::{
+    executable::{Executable, ExecutableDescriptor},
+    write_set::TransactionWrite,
+};
+use crossbeam::utils::CachePadded;
+use dashmap::DashMap;
+use std::{
+    collections::{btree_map::BTreeMap, HashMap},
+    hash::Hash,
+    sync::{
+        atomic::{AtomicUsize, Ordering},
+        Arc,
+    },
+};
+
+const FLAG_DONE: usize = 0;
+const FLAG_ESTIMATE: usize = 1;
+
+/// Every entry in shared multi-version data-structure has an "estimate" flag
+/// and some content.
+struct Entry<V: TransactionWrite> {
+    /// Used to mark the entry as a "write estimate". Even though the entry
+    /// lives inside the DashMap and the entry access will have barriers, we
+    /// still make the flag Atomic to provide acq/rel semantics on its own.
+    flag: AtomicUsize,
+
+    /// The contents of the module as produced by the VM (can be WriteOp based on a
+    /// blob or CompiledModule, but must satisfy TransactionWrite to be able to
+    /// generate the hash below.
+    module: Arc<V>,
```

Comment:
We do not store the executable here?
Is it because the executable is a function of other modules?

Also, if during validation hash is the same but the executable is different because dependencies changed? 
Will I abort when fetching dependencies? What if I do not need them? Then we are fine?

Reply:
we don't care about dependencies for now.
executable is a function of module, different version of VM may produce different executable based on the same module (code). module is code.